test/activerecord/cte_test.rb

```diff
@@ -142,4 +142,25 @@ def test_with_when_merging_relations
     assert_equal merged.size, 1
     assert_equal merged[0].views_count, 123
   end
+
+  def test_with_when_merging_relations_with_identical_with_names_and_identical_quries
+    most_popular1 = Post.with(most_popular: Post.where("views_count >= 100"))
+    most_popular2 = Post.with(most_popular: Post.where("views_count >= 100"))
```

Comment:
These still fail since the equality check isn't doing any sort of conversion. This would require some more extensive rejiggering I think.

Reply:
@vlado @joelvh in the case of Rails 6.1 and using strings as `#with` values, this broke in Rails 6.1 due to this commit:

https://github.com/rails/rails/commit/596469d556fa924bd792288668f7f0fbdea518f2

There's no detection for strings in there. Before they just slid on through, but now they need to be accounted for.

I have a patch that I'll submit upstream, just wrapping that up now. As for what to do for the gem, a monkey patch would probably be the only recourse, as the code in question is in the guts of the SQL generation visitor methods.

Reply:
Rails PR is up at https://github.com/rails/rails/pull/42563, with an issue report at https://github.com/rails/rails/issues/42561.

Reply:
> These still fail since the equality check isn't doing any sort of conversion. This would require some more extensive rejiggering I think.

I suggest we skip that then. Let's merge this one and open new PR when/if needed.

Reply:
> I have a patch that I'll submit upstream, just wrapping that up now. As for what to do for the gem, a monkey patch would probably be the only recourse, as the code in question is in the guts of the SQL generation visitor methods.

Great find and great that you submitted PR to the Rails 👍 Let's see when it gets merged and released.

I would even be ok if we just add a note to the Readme that passing string for some Rails versions does not work due to the mentioned bug.

Reply:
Yeah, that failed in my tests w/ 6.1 and i ignored it because I'm not using strings 😬 

Can we use the SQL literal suggested in your upstream PR internally here?